pkg/apis/kops/cluster.go

```diff
@@ -158,6 +158,9 @@ type ClusterSpec struct {
 	CloudConfig                    *CloudConfiguration           `json:"cloudConfig,omitempty"`
 	ExternalDNS                    *ExternalDNSConfig            `json:"externalDns,omitempty"`
 
+	// Components contains the configuration for kops components.
+	Components *ComponentConfig `json:"components,omitempty"`
```

Comment:
What is the definition of a "component"? What determines if something is a "component" versus an "addon" or just has its own subfield in the cluster spec?

Reply:
The reason for adding a new struct was to not further "bloat" the cluster spec. Was briefly mentioned on slack.

Addons was already used for those where you provide manifest urls and such. Didn't quite seem to quite fit what I wanted to do.
The reason for calling it component is that we the API spec for this is in componentconfig.go and that the options builders are also called components.

If this part goes in, I want to move CAS in there as well. In future spec versions, other components like kubedns, kubeproxy etc could go in there as well.

Reply:
I am ok with having a new structure to keep new addons separated. Depending on the agreed purpose, the name is not quite right.
I don't think kubedns, kubeproxy, ... can be moved here, because the have some established sections already.

Reply:
Just to be clear: I am thinking of when we do a full new API version.

Naming things is hard, but it was the only one I could think of besides "addons" that was already in use.

Reply:
I agree about the "hard" part :). I would see kube-proxy, kube-... in an Kubernetes struct, not in the "addons" one that should be optional managed things.

Reply:
So I could always add this to the addons struct instead, but it does feel a bit strange to me.

Reply:
Adding a new level to the v1alpha2 API makes it even more inconsistent than it already is. It might make sense to reorganize things when we can do a v1beta1 API, but adding a new level at this point is not going to make things any less confusing.

This is fundamentally an AWS-specific feature and should be off of the AWS-specific cloud provider settings. The fact that it is implemented as a bootstrap addon is an implementation detail.

Reply:
Sounds good to me. So, same way as autoscaler was done.

Reply:
Moving it to the AWS struct sounds fine for me. But autoscaler stays where it is then.